File: app/src/main/java/com/linxiao/quickdevframework/sample/adapter/HeaderFooterActivity.java
package com.linxiao.quickdevframework.sample.adapter;

import android.os.Bundle;

import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.chad.library.adapter4.QuickAdapterHelper;
import com.linxiao.framework.architecture.SimpleViewBindingActivity;
import com.linxiao.framework.list.HeaderFooterAdapter;
import com.linxiao.quickdevframework.R;
import com.linxiao.quickdevframework.databinding.ActivityHeaderFooterBinding;

import java.util.Arrays;

import kotlin.Unit;

public class HeaderFooterActivity extends SimpleViewBindingActivity<ActivityHeaderFooterBinding> {

    private SampleAdapter mAdapter;
    private QuickAdapterHelper quickAdapterHelper;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mAdapter = new SampleAdapter();
        mAdapter.setOnItemClickListener((baseQuickAdapter, view, i) -> {
            addHeaderView();
            addFooterView();
        });
        quickAdapterHelper = new QuickAdapterHelper.Builder(mAdapter)
                .build();
        getViewBinding().rcvHeaderFooter.setLayoutManager(new LinearLayoutManager(this));
        getViewBinding().rcvHeaderFooter.setItemAnimator(new DefaultItemAnimator());
        getViewBinding().rcvHeaderFooter.setAdapter(quickAdapterHelper.getAdapter());
        HeaderFooterAdapter headerAdapter = new HeaderFooterAdapter(R.layout.header_sample);
        headerAdapter.setOnInitView(quickViewHolder -> {
            quickViewHolder.itemView.setOnClickListener(v -> {
                addHeaderView();
            });
            return Unit.INSTANCE;
        });
        quickAdapterHelper.addBeforeAdapter(headerAdapter);

        HeaderFooterAdapter footerAdapter = new HeaderFooterAdapter(R.layout.footer_sample);
        footerAdapter.setOnInitView(quickViewHolder -> {
            quickViewHolder.itemView.setOnClickListener(v -> {
                addFooterView();
            });
            return Unit.INSTANCE;
        });
        quickAdapterHelper.addAfterAdapter(footerAdapter);

        initData();
    }

    private void initData() {
        mAdapter.submitList(Arrays.asList("1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"));
    }

    private void addHeaderView() {
        HeaderFooterAdapter headerAdapter = new HeaderFooterAdapter(R.layout.header_added);
        headerAdapter.setOnInitView(quickViewHolder -> {
            quickViewHolder.itemView.setOnClickListener(v -> {
                quickAdapterHelper.removeAdapter(headerAdapter);
            });
            return Unit.INSTANCE;
        });
        quickAdapterHelper.addBeforeAdapter(0, headerAdapter);
    }

    private void addFooterView() {
        HeaderFooterAdapter footerAdapter = new HeaderFooterAdapter(R.layout.footer_added);
        footerAdapter.setOnInitView(quickViewHolder -> {
            quickViewHolder.itemView.setOnClickListener(v -> {
                quickAdapterHelper.removeAdapter(footerAdapter);
            });
            return Unit.INSTANCE;
        });
        quickAdapterHelper.addAfterAdapter(footerAdapter);
    }
}

File: app/src/main/java/com/linxiao/quickdevframework/sample/adapter/HeaderFooterActivity.kt
package com.linxiao.quickdevframework.sample.adapter

import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import com.chad.library.adapter4.BaseQuickAdapter
import com.chad.library.adapter4.QuickAdapterHelper
import com.chad.library.adapter4.viewholder.QuickViewHolder
import com.linxiao.framework.architecture.SimpleViewBindingActivity
import com.linxiao.framework.list.HeaderFooterAdapter
import com.linxiao.quickdevframework.R
import com.linxiao.quickdevframework.databinding.ActivityHeaderFooterBinding

class HeaderFooterActivity : SimpleViewBindingActivity<ActivityHeaderFooterBinding>() {
    
    private val mAdapter: SampleAdapter by lazy {
        SampleAdapter()
    }
    
    private val quickAdapterHelper: QuickAdapterHelper by lazy {
        QuickAdapterHelper.Builder(mAdapter).build()
    }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mAdapter.setOnItemClickListener { baseQuickAdapter: BaseQuickAdapter<String, *>, view: View, i: Int ->
            addHeaderView()
            addFooterView()
        }
        viewBinding.rcvHeaderFooter.setLayoutManager(LinearLayoutManager(this))
        viewBinding.rcvHeaderFooter.setItemAnimator(DefaultItemAnimator())
        viewBinding.rcvHeaderFooter.setAdapter(quickAdapterHelper.adapter)
        val headerAdapter = HeaderFooterAdapter(R.layout.header_sample)
        headerAdapter.onInitView = { quickViewHolder: QuickViewHolder ->
            quickViewHolder.itemView.setOnClickListener { v: View? -> addHeaderView() }
            Unit
        }
        quickAdapterHelper.addBeforeAdapter(headerAdapter)
        val footerAdapter = HeaderFooterAdapter(R.layout.footer_sample)
        footerAdapter.onInitView = { quickViewHolder: QuickViewHolder ->
            quickViewHolder.itemView.setOnClickListener { v: View? -> addFooterView() }
            Unit
        }
        quickAdapterHelper.addAfterAdapter(footerAdapter)
        initData()
    }

    private fun initData() {
        mAdapter.submitList(
            mutableListOf(
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1",
                "1"
            )
        )
    }

    private fun addHeaderView() {
        val headerAdapter = HeaderFooterAdapter(R.layout.header_added)
        headerAdapter.onInitView = { quickViewHolder: QuickViewHolder ->
            quickViewHolder.itemView.setOnClickListener { v: View? ->
                quickAdapterHelper.removeAdapter(
                    headerAdapter
                )
            }
            Unit
        }
        quickAdapterHelper.addBeforeAdapter(0, headerAdapter)
    }

    private fun addFooterView() {
        val footerAdapter = HeaderFooterAdapter(R.layout.footer_added)
        footerAdapter.onInitView = { quickViewHolder: QuickViewHolder ->
            quickViewHolder.itemView.setOnClickListener { v: View? ->
                quickAdapterHelper.removeAdapter(
                    footerAdapter
                )
            }
            Unit
        }
        quickAdapterHelper.addAfterAdapter(footerAdapter)
    }
}
